_practicalities/intro.md

```diff
@@ -0,0 +1,355 @@
+
+See the following url for more up to date informations:
+
+https://i.am.an/url
+
+""")
+
+```
+
+## Watch out for beta releases
+
+
+Make sure your version number match pep 440 or you will get surprises during
+beta in particular as the `sdist` and `wheel` will appear as being different
+versions, in particular sdist (during beta/rc/post) can appear with a greater
+version number than wheels. Pip thus try to install the sdist instead of the
+wheel, which have more chance of failing, in particular with pre 24.2 versions
+of setuptools.
+
+The regular expression to check for validity of pep440 can be find below:
+
+  `^([1-9]\\d*!)?(0|[1-9]\\d*)(\\.(0|[1-9]\\d*))*((a|b|rc)(0|[1-9]\\d*))?(\\.post(0|[1-9]\\d*))?(\\.dev(0|[1-9]\\d*))?`
+
+## Depend on setuptools
+
+You can mark your library as dependent on setuptools greater than 24.3 starting
+now, this will insure that during the next upgrade (when the packages drop
+python 2 support) will have the right version of setuptools.
+
+Of course regardless of all the care you will take for your library to no break
+and to install only on python 2, you will likely have cases where it still end
+up being installed on incompatible versions of Python. Simply because users
+upgrades rarely and only an old version of pip or setuptools is enough to make
+the all update process broken.
+
+## fail early in setup.py
+
+Leave `setup.py` python 2 compatible and fail early. If you detect Python 2
+raise a clear error message and ask user to make sure they have pip >9.0 (or
+migrate to Python 3). You can (try to) conditionally import pip and check for
+its version but this might not be the same pip. Failing early is important to
+make sure the Python installation does not install and incompatible version.
+Otherwise user code can fail at runtime arbitrary later in the future, which can
+be a difficult to debug and fix. Get inspiration from the message of failure at
+runtime, and adapt for installation time.
+
+## Fix dependant libraries
+
+If you control dependant packages, Make sure to include conditional dependencies
+depending on the version of Python.
+
+# Incorrect mitigations
+
+This is a collection of "mitigation" or "solutions" you will find on the web
+and that you will hear about. This is an attempt to acknowledge them, and
+explain why they can't work and what are their drawbacks before you attempt to
+implement them.
+
+### Use a meta-package.
+
+It is possible to release a meta-package that has _virtually_ no code and rely
+on conditional dependency to install its actual core code on the user system.
+For example, Frob-6.0 could be a meta-package which depends on
+Frob-real-py2 on Python <3.0, and Frob-real-py3 on Python >= 3.4. While
+this approach is _doable_ this can make imports confusing.
+
+Moreover, upgrading your package may need the user to explicitly tell pip to
+upgrade dependencies as `pip install -U frob` will only upgrade the meta-package.
```

Comment:
Eliding detail is fine so long as it's not misleading, but "will only upgrade the meta-package" is inaccurate - if it updates the metapackage, it will update the real package too. The practical upshot for package authors is that you need to make a release of the metapackage whenever you make a release of the real package.